### PR TITLE
admin: support /drain_listeners?skip_exit

### PR DESCRIPTION
**Pull request by howardjohn**
Commit Message: `admin: support /drain_listeners?skip_exit`
Additional Description:
Due to changes in Kubernetes (https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) around sidecars, there is a more optimal way to manage the lifecycle of Envoy as a sidecar in Kubernetes. However, it is not currently possible with the current knobs.

Our goal:
1. When the pod starts terminating, start "draining" inbound connections. Accept new ones, but respond with GOAWAY/connection=close to discourage clients
2. In parallel (outside of envoy), clients should be discourage to making new connections _generally_
3. Once the application container shuts down, we have nothing left to do; as a sidecar proxy, we exist only to serve the application container. Shutdown ~immediately

The time between (1) and (3) can be quite large (but not unbounded; Kubernetes has a max time before it will SIGKILL everything).

How we want to implement this:

1. When pod starts terminating, `/drain_listeners?graceful&skip_exit`
2. When application container shutsdown, SIGTERM Envoy.

The reason we need a `skip_exit` is due to the timing between the two. If we just do this without `skip_exit` if the application container takes > `drain-time-s` we will block new connections unexpectedly. While we could just raise the `drain-time`, this also impacts other draining (notably, LDS update), which may be unexpected.

Risk Level: Low
Testing: Added new integration test based on existing ones around draining
Docs Changes: Add docs to admin.rst

### CODE REVIEW

File: source/server/admin/listeners_handler.cc [hunk 1]
```diff
@@ -22,12 +22,15 @@ Http::Code ListenersHandler::handlerDrainListeners(Http::ResponseHeaderMap&,
                                                  : ListenerManager::StopListenersType::All;
 
   const bool graceful = params.find("graceful") != params.end();
+  const bool skip_exit = params.find("skip_exit") != params.end();
```
Comment (soulxu) [inline source/server/admin/listeners_handler.cc:25]: It seems we should add new param here also https://github.com/envoyproxy/envoy/blob/a5930b27b083d71a92ca66ec5200cca65721f0a7/source/server/admin/admin.cc#L200

File: source/server/admin/listeners_handler.cc [hunk 2]
```diff
@@ -22,12 +22,15 @@ Http::Code ListenersHandler::handlerDrainListeners(Http::ResponseHeaderMap&,
...
   if (graceful) {
     // Ignore calls to /drain_listeners?graceful if the drain sequence has
     // already started.
     if (!server_.drainManager().draining()) {
-      server_.drainManager().startDrainSequence([this, stop_listeners_type]() {
-        server_.listenerManager().stopListeners(stop_listeners_type, {});
+      server_.drainManager().startDrainSequence([this, stop_listeners_type, skip_exit]() {
+        if (!skip_exit) {
```
Comment (soulxu) [inline source/server/admin/listeners_handler.cc:31]: We can return BadRequest if skip_exit is set but no graceful

**Review by soulxu — CHANGES_REQUESTED** [review]

File: docs/root/operations/admin.rst [hunk 3]
```diff
@@ -350,6 +350,10 @@ modify different aspects of the server:
    This behaviour and duration is configurable via server options or CLI
    (:option:`--drain-time-s` and :option:`--drain-strategy`).
 
+   .. http:post:: /drain_listeners?graceful&skip_exit
+
+   When draining listeners, do not exit after the drain period. This must be used with `graceful`.
```
Comment (soulxu) [inline docs/root/operations/admin.rst:355]: We better to have an item for changelog https://github.com/envoyproxy/envoy/blob/a5930b27b083d71a92ca66ec5200cca65721f0a7/changelogs/current.yaml#L200

**Review by soulxu — DISMISSED** [review]
LGTM except main merge needed, thanks

File: changelogs/current.yaml [hunk 4]
```diff
@@ -386,6 +386,10 @@ new_features:
 - area: geoip
   change: |
     Added support for :ref:`Maxmind geolocation provider <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig>`.
+    added support for the internal listener address recovery using the original destination listener filter.
```
Comment (kyessenov) [inline changelogs/current.yaml:389]: Bad merge? Can you remove this line?

**Review by kyessenov — CHANGES_REQUESTED** [review]
LGTM, please update the change notes.

File: changelogs/current.yaml (hunk 4 above)
Comment (howardjohn) [reply changelogs/current.yaml]: thanks! done

**Review by kyessenov — DISMISSED** [review]

**Comment by howardjohn** [comment]
> @howardjohn you need to fix the format. Looks like some white space issue by your editor. Can you run `bazel run //tools/code_format:check_format -- fix` and apply the fixes?

Ok I don;t really understand why me adding 1 line means the entire file should reformat, but done. Hope that works

File: changelogs/current.yaml [hunk 5]
```diff
@@ -386,6 +386,9 @@ new_features:
 - area: geoip
   change: |
     Added support for :ref:`Maxmind geolocation provider <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig>`.
+- area: admin
+  change: |
+    Added a new `skip_exit` query parameter to `/drain_listeners` to skip exiting after the drain period.
```
Comment (kyessenov) [inline changelogs/current.yaml:391]: Double back-ticks, please, for `skip_exit` and `drain_listeners`.

File: source/server/admin/admin.cc [hunk 6]
```diff
@@ -124,7 +124,7 @@ AdminImpl::AdminImpl(const std::string& profile_path, Server::Instance& server,
       runtime_handler_(server), listeners_handler_(server), server_cmd_handler_(server),
       server_info_handler_(server),
       // TODO(jsedgwick) add /runtime_reset endpoint that removes all admin-set values
-      handlers_{
+      handlers_ {
```
Comment (kyessenov) [inline source/server/admin/admin.cc:127]: Looks like a true compilation problem here?

Comment (howardjohn) [reply source/server/admin/admin.cc]: I think I finally got it... thanks for the patience :slightly_smiling_face:

**Comment by kyessenov** [comment]
Can you resolve the conflict? Also, please don't force push.

**Review by kyessenov — APPROVED** [review]

**Pull request merged**